From 80fdce7f93a32e47789a175ee646637987635a87 Mon Sep 17 00:00:00 2001
From: John Howard <howardjohn@google.com>
Date: Tue, 2 Nov 2021 17:32:51 -0700
Subject: Add DeepCopy to echo CallOptions (#35848)

When using nodeport, we mutate the Calloptions port (a pointer). This
means on subsequent retries things get messed up.

Tested in https://github.com/istio/istio/pull/35354
---
 pkg/test/framework/components/echo/call.go     | 14 ++++++++++++++
 pkg/test/framework/components/istio/ingress.go |  1 +
 2 files changed, 15 insertions(+)

diff --git a/pkg/test/framework/components/echo/call.go b/pkg/test/framework/components/echo/call.go
index 078581ef04..fdaf8d21d3 100644
--- a/pkg/test/framework/components/echo/call.go
+++ b/pkg/test/framework/components/echo/call.go
@@ -123,6 +123,20 @@ func (o CallOptions) GetHost() string {
 	return ""
 }
 
+func (o CallOptions) DeepCopy() CallOptions {
+	clone := o
+	if o.Port != nil {
+		dc := *o.Port
+		clone.Port = &dc
+	}
+	if o.Alpn != nil {
+		clone.Alpn = &proto.Alpn{
+			Value: o.Alpn.GetValue(),
+		}
+	}
+	return clone
+}
+
 // Validator validates that the given responses are expected.
 type Validator interface {
 	// Validate performs the validation check for this Validator.
diff --git a/pkg/test/framework/components/istio/ingress.go b/pkg/test/framework/components/istio/ingress.go
index 6d05f8567e..6c6d1075df 100644
--- a/pkg/test/framework/components/istio/ingress.go
+++ b/pkg/test/framework/components/istio/ingress.go
@@ -204,6 +204,7 @@ func (c *ingressImpl) callEcho(options echo.CallOptions, retry bool, retryOption
 		addr string
 		port int
 	)
+	options = options.DeepCopy()
 	if options.Port.ServicePort == 0 {
 		// Default port based on protocol
 		switch options.Port.Protocol {
-- 
2.35.3

